### PR TITLE
Added select all checkbox to map

### PR DESCRIPTION
**Pull request by reddy1praveen**

### CODE REVIEW

File: app/frontend/js/map.js [hunk 1]
```diff
@@ -110,12 +109,25 @@ const convertToParcelSheetId = (parcelId) => {
 
 const addCheckboxEventListener = (checkbox, selectfeatures, parcelSource) => {
   checkbox.addEventListener('change', (e) => {
-    const parcelId = convertToParcelSheetId(checkbox.id)
+    //console.info('Parcel parcelSource', checkbox.id)
     const parcelFeatures = parcelSource.getFeatures()
-    for (const feature of parcelFeatures) {
-      if (feature.get('parcel_id') === parcelId[1] && feature.get('sheet_id') === parcelId[0]) {
+    if(checkbox.id === "SelectAll") {
+      for (const feature of parcelFeatures) {
```
Comment (steve-dickinson) [inline app/frontend/js/map.js:115]: Could this be in its own function. So split check all and check individual into their own functions

File: app/routes/funding-options/models/how-much.js [hunk 2]
```diff
@@ -73,8 +75,20 @@ const getAllItems = (selectedStandard, selectedParcels) => {
         ? selectedParcels.find(item => item.id === x.id).value
         : Number(x.area).toFixed(2),
       warnings: []
-    }
+    } 
   ))
+  const NewCheckBoxItem  =  {
```
Comment (steve-dickinson) [inline app/routes/funding-options/models/how-much.js:80]: Rename to describe what the elements purpose is, so maybe selectAllCheckBoxItem

File: app/frontend/js/map.js [hunk 3]
```diff
@@ -110,12 +109,25 @@ const convertToParcelSheetId = (parcelId) => {
 
 const addCheckboxEventListener = (checkbox, selectfeatures, parcelSource) => {
   checkbox.addEventListener('change', (e) => {
-    const parcelId = convertToParcelSheetId(checkbox.id)
+    //console.info('Parcel parcelSource', checkbox.id)
```
Comment (steve-dickinson) [inline app/frontend/js/map.js:112]: Remove commented out console.log

File: app/routes/funding-options/models/how-much.js [hunk 4]
```diff
@@ -73,8 +75,20 @@ const getAllItems = (selectedStandard, selectedParcels) => {
...
+    "text": "Select All",
+    "value": "SelectAll",
+    "hint": {
```
Comment (steve-dickinson) [inline app/routes/funding-options/models/how-much.js:83]: Doesn't need the hint text

**Review by steve-dickinson — CHANGES_REQUESTED** [review]
The select all checkbox isn't necessarily going to be part of the parcel check boxes on page design. So it would be beneficial if the select all could be an individual checkbox. This will give us the freedom to relocate it within the page design , if required.

File: app/frontend/js/map.js [hunk 5]
```diff
@@ -110,12 +109,25 @@ const convertToParcelSheetId = (parcelId) => {
...
+      for (const feature of parcelFeatures) {
         e.target.checked ? selectfeatures.push(feature) : selectfeatures.remove(feature)
       }
+      const checkBoxes = document.getElementsByClassName('govuk-checkboxes__input')
+      const selectAllParcelCheckBox = document.getElementById("SelectAll")
+      for (const checkbox of checkBoxes) {
+        const parcelCheckBox = document.getElementById(checkbox.id)
+        parcelCheckBox.checked = selectAllParcelCheckBox.checked
+      }
+    } else  {
```
Comment (steve-dickinson) [inline app/frontend/js/map.js:124]: need to remove trailing space after else

**Review by steve-dickinson — CHANGES_REQUESTED** [review]
You need to run scripts/test as there are a few linting errors found

**Review by steve-dickinson — COMMENTED** [review]
I think you could simplify it.

If you add a check box group to the `how-much.template.njk`

```
<div class="govuk-checkboxes" data-module="govuk-checkboxes">
  <div class="govuk-checkboxes__item">
    <input class="govuk-checkboxes__input" id="selectAllParcels" name="selectAllParcels" type="checkbox" value="all">
    <label class="govuk-label govuk-checkboxes__label" for="selectAllParcels">
      Select all
    </label>
  </div>
</div>
```
Then within `map.js` add a `selectAllParcels` function:

```
const selectAllParcels = (selectfeatures, parcelSource) => {
  const selectAll = document.getElementById('selectAllParcels')
  selectAll.addEventListener('change', () => {
    const checkBoxes = document.getElementsByClassName('govuk-checkboxes__input')
    for (const checkbox of checkBoxes) {
      checkbox.checked = selectAll.checked
      addToSelectFeatures(selectfeatures, parcelSource, checkbox, checkbox.id)
    }
  })
}
```

So this sets up a listener on the selectAll checkbox and when checked updates all the parcel check boxes and the selected features

Add a new function `addToSelectFeatures `:

```
const addToSelectFeatures = (selectfeatures, parcelSource, target, id) => {
  const parcelId = convertToParcelSheetId(target.id)
  const parcelFeatures = parcelSource.getFeatures()
  for (const feature of parcelFeatures) {
    if (feature.get('parcel_id') === parcelId[1] && feature.get('sheet_id') === parcelId[0]) {
      target.checked ? selectfeatures.push(feature) : selectfeatures.remove(feature)
    }
  }
}
```

Change the `addCheckboxEventListener` to use the `addToSelectFeatures ` instead:

```
const addCheckboxEventListener = (checkbox, selectfeatures, parcelSource) => {
  checkbox.addEventListener('change', (e) => {
    addToSelectFeatures(selectfeatures, parcelSource, e.target, e.target.id)
  })
}
```

Then just add `selectAllParcels(selectfeatures, parcelSource)` to the `parcelSelection` function

```
const parcelSelection = (map, allowSelect, selectedParcels, parcelSource) => {
  if (allowSelect) {
    const selectClick = selectLayer(map)
    const selectfeatures = selectClick.getFeatures()
    checkBoxSelection(parcelSource, selectfeatures)
    selectPointerMove(map)
    preParcelSelection(selectedParcels, parcelSource, selectfeatures)
    selectAllParcels(selectfeatures, parcelSource)
  }
}

```
There is probabally some refactoring you can do afterwards with `selectAllParcels`, `addCheckboxEventListener` and `checkBoxSelection`

**Comment by reddy1praveen** [comment]
closing  this PR as opened another one as per Steve Code change suggestions : https://github.com/DEFRA/ffc-sfi-apply-web/pull/110

**Pull request closed without merging**